openhtf/core/test_executor.py

```diff
@@ -115,45 +124,64 @@ def _thread_proc(self):
       self.test_state = test_state.TestState(self._test_descriptor, self.uid)
       phase_exec = phase_executor.PhaseExecutor(self.test_state)
 
-      # Any access to self._exit_stack must be done while holding this lock.
+      # Any access to self._exit_stacks must be done while holding this lock.
       with self._lock:
         self._exit_stack = exit_stack
-        # We need to exit the PhaseExecutor before we tear down the plugs
-        # that a phase may be executing, so we add them in reverse order.
-        exit_stack.callback(self.test_state.plug_manager.tear_down_plugs)
-        exit_stack.callback(phase_exec.stop)
+        # Ensure that we tear everything down when exiting.
+        exit_stack.callback(self._execute_test_teardown, phase_exec)
+        # We don't want to run the 'teardown function' unless the test has
+        # actually started.
+        self._do_teardown_function = False
 
-      # Have the phase executor run the start trigger phase. Do partial plug
-      # initialization for just the plugs needed by the start trigger phase.
       if self._test_start is not None:
-        self.test_state.plug_manager.initialize_plugs(
-            (phase_plug.cls for phase_plug in self._test_start.plugs))
-        phase_exec.execute_start_trigger(self._test_start)
+        self._execute_test_start(phase_exec)
+      # The trigger has run and the test has started, so from now on we want the
+      # teardown function to execute at the end, no matter what.
+      with self._lock:
+        self._do_teardown_function = True
       self.test_state.mark_test_started()
 
       # Full plug initialization happens _after_ the start trigger, as close to
       # test execution as possible, for the best chance of test equipment being
       # in a known-good state at the start of test execution.
       self.test_state.plug_manager.initialize_plugs()
 
-      # Everything is set, set status and begin test execution.  Note we don't
-      # protect this with a try: block because the PhaseExecutor handles any
-      # exceptions from test code.  Any exceptions here are caused by the
-      # framework, and we probably want them to interrupt framework state
-      # changes (like the transition to FINISHING).
+      # Everything is set, set status and begin test execution.
       self._execute_test_phases(phase_exec)
 
+  def _execute_test_start(self, phase_exec):
```

Comment:
This name (and even the docstring) kind of hides that this is also where plug initialization happens.  Maybe pull just that first line out and keep it above at line 137?  I think the plug init -> start phase -> test phases -> finalize flow should be visible up there.